neuralpp/symbolic/basic_expression.py

```diff
@@ -167,8 +173,53 @@ def number_of_arguments(self) -> int:
 
     def syntactic_eq(self, other) -> bool:
         match other:
-            case BasicFunctionApplication(function=function, arguments=arguments, type=other_type):
-                return all(lhs.syntactic_eq(rhs) for lhs, rhs in zip(self.subexpressions, [function] + arguments)) and \
-                       self.type == other_type
+            case BasicFunctionApplication(subexpressions=other_subexpressions):
```

Comment:
We don't need to compare type since the equivalence of `function` implies the equivalence of type of the function application, if the argument numbers are equal.